aztarna/ros/ros2/scanner.py

```diff
@@ -0,0 +1,170 @@
+            for node in host.nodes:
+                print(f'\t\tNode Name: {node.name} \t|\t Namespace: {node.namespace}')
+                if self.extended:
+                    self.print_node_topics(node)
+            print('-' * 80)
+
+    @staticmethod
+    def print_node_topics(node):
+        """
+        Helper function for printing node related topics only.
+
+        :param node: Node containing the topics.
+        """
+        print(f'\t\tPublished topics:')
+        for topic in node.published_topics:
+            print(f'\t\t\tTopic Name: {topic.name} \t|\t Topic Type: {topic.topic_type}')
+        print('\t\tSubscribed topics:')
+        for topic in node.subscribed_topics:
+            print(f'\t\t\tTopic Name: {topic.name} \t|\t Topic Type: {topic.topic_type}')
+
+    def write_to_file(self, out_file):
+        """
+        Write scanner results to the specified output file.
+
+        :param out_file: Output file to write the results on.
+        """
+        lines = []
+        header = 'DomainID;NodeName;Namespace;Topic;TopicType;Direction\n'
+        lines.append(header)
+        with open(out_file, 'w') as f:
+            for host in self.found_hosts:
+                if self.extended:
+                    for node in host.nodes:
+                        self.write_node_topics(host, lines, node)
+                else:
+                    for topic in host.topics:
+                        line = f'{host.domain_id};;{topic.name};{topic.topic_type};;\n'
+                        lines.append(line)
+            f.writelines(lines)
+
+    @staticmethod
+    def write_node_topics(host, lines, node):
+        """
+        Helper function to generate the node related topic information lines for writing in file.
+
+        :param host: :class:`aztarna.ros.ros2.helpers.ROS2Host` class object.
+        :param lines: list containing the lines to write on the file.
+        :param node: :class:`aztarna,.ros.ros2.helpers.ROS2Node` class object containing the information to write.
+        """
+        for published_topic in node.published_topics:
+            line = f'{host.domain_id};{node.name};{node.namespace};{published_topic.name};' \
+                f'{published_topic.topic_type};Publish\n'
+            lines.append(line)
+        for subscribed_topic in node.subscribed_topics:
+            line = f'{host.domain_id};{node.name};{node.namespace};{subscribed_topic.name};' \
+                f'{subscribed_topic.topic_type};Subscribe\n'
+            lines.append(line)
+
+    def scan(self):
+        """
+        Scan the local network and all available ROS_DOMAIN_IDs against ROS2 nodes.
+
+        :return: A list containing the found ROS2 systems.
+        """
+        try:
+            import rclpy
+        except ImportError:
+            raise Exception('ROS2 needs to be installed and sourced to run ROS2 scans')
+
+        for i in range(0, max_ros_domain_id):
+            os.environ['ROS_DOMAIN_ID'] = str(i)
```

Comment:
What about implement the same for different RMW implementation?

Reply:
Shouldn't they be interoperable between different middlewares? At least to some basic level

Reply:
Not in all ROS2 versions AFAIK.

Reply:
We are attempting to create a method to detect the installed RMW implementations, but without success. The first idea was to alter the RMW_IMPLEMENTATION environment variable prior to calling rclpy.init, in a try-except clause, but rclpy finishes the process on error.
Any idea on how we could implement it?? Listing installed packages maybe?

Reply:
Let me have a look.

Reply:
@LanderU , I have created a preliminary version of the scanner, but I'm getting opensplice and connext related errors when running it. Let's see if we can find the cause.
![image](https://user-images.githubusercontent.com/1711272/59699919-ae116700-91f2-11e9-9422-05f99e48be6d.png)

Reply:
Humm, I'll have a look into this this night.

Reply:
```bash
[+] Host found in Domain ID 40
	Topics:
	Nodes:
		Node Name: _ros2cli_daemon_40 	|	 Namespace: /
		Node Name: talker 	|	 Namespace: /
--------------------------------------------------------------------------------
[+] Host found in Domain ID 40
	Topics:
	Nodes:
		Node Name: _ros2cli_daemon_40 	|	 Namespace: /
		Node Name: talker 	|	 Namespace: /
--------------------------------------------------------------------------------
[+] Host found in Domain ID 40
	Topics:
	Nodes:
		Node Name: _ros2cli_daemon_40 	|	 Namespace: /
		Node Name: talker 	|	 Namespace: /
--------------------------------------------------------------------------------
```

Using OpenSplice.

Reply:
I'm going to try to create a patch.

Reply:
```bash
========================================================================================
Context     : DDS::DomainParticipantFactory::create_participant
Date        : 2019-06-18T19:50:19+0000
Node        : 3e4a113cc568
Process     : python3.6 <9951>
Thread      : main thread 7fd60e17a740
Internals   : DomainParticipantFactory.cpp/324/6.9.190403OSS///-1
----------------------------------------------------------------------------------------
Report      : Unable to connect to domain id = 1.
              The most common causes of this error are an incorrect configuration file or
              that OpenSpliceDDS is not running (when using shared memory mode).
Internals   : u_participantNew/u_participant.c/234/773/1560887419.960899293
----------------------------------------------------------------------------------------
Report      : Failed to start Spliced for domain 'ospl_sp_ddsi' within 0 seconds, result = U_RESULT_PRECONDITION_NOT_MET

Internals   : user::u_domain::startSpliceThread/u_domain.c/1253/781/1560887419.960924533

>>> [rcutils|error_handling.c:106] rcutils_set_error_state()
This error state is being overwritten:

  'failed to create domain participant, at /tmp/binarydeb/ros-dashing-rmw-opensplice-cpp-0.7.1/src/rmw_node.cpp:162, at /tmp/binarydeb/ros-dashing-rcl-0.7.4/src/rcl/node.c:326'

with this new error message:

  'rcl node's rmw handle is invalid, at /tmp/binarydeb/ros-dashing-rcl-0.7.4/src/rcl/node.c:464'

rcutils_reset_error() should be called after error handling to avoid this.
<<<
[ERROR] [rcl]: Failed to fini publisher for node: 1
```

Reply:
https://istkb.adlinktech.com/article/configuring-shared-memory/